File: capacityToShipPackages.py
class Solution:
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        left = max(weights)
        right = 0
        
        for i in range(len(weights)):
            right += weights[i]
            
     
        # left = leastCapacity
        # right = maxCapacity
        
        while left <= right:
            #mid is the capacity
            mid = left + (right-left)//2
            currSum = 0
            days = 1
            for i in range(len(weights)):
                if currSum + weights[i] > mid:
                    currSum = 0
                    days += 1
                currSum += weights[i]
                
            if days > D:
                left = mid + 1
            else:
                right = mid-1
                
        return left

File: CapacityToShipPackages.java
//TC : O(Nlog(Sum of weights - maxWeight)
//SC : O(1)
class CapacityToShipPackages {
    public int shipWithinDays(int[] weights, int days) {

        int minCapacity = 0;
        int maxCapacity = 0;

        for(int weight : weights){
            minCapacity = Math.max(minCapacity, weight);
            maxCapacity = maxCapacity + weight;
        }

        int start = minCapacity;
        int end = maxCapacity;



        while(start <= end){
            int mid = start + (end - start)/2;
            int daysNeeded = 1; int currentWeight = 0;
            for(int weight : weights){
                if(currentWeight + weight > mid){
                    daysNeeded++;
                    currentWeight = 0;
                }
                currentWeight+=weight;
            }

            if(daysNeeded > days){
                start = mid +1;
            }
            else {
                end = mid -1;
            }

        }
        return start;
    }
}